### PR TITLE
Skip explanatory text at the start of cert byte blobs if we find a PEM header.

### PR DESCRIPTION
**Pull request by jkoritzinsky**
This enables support for PEM-formatted certs prepended with explanatory text.

Contributes to #45741

### CODE REVIEW

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c [hunk 1]
```diff
@@ -287,3 +295,52 @@ static int32_t PopulateByteArray(JNIEnv* env, jbyteArray source, uint8_t* dest,
...
+static void FindCertStart(const uint8_t** buffer, int32_t* len)
+{
+    if (iscntrl(**buffer) && !isspace(**buffer))
+    {
+        // If the character is a control character that isn't whitespace, then we're probably using a DER encoding
+        // and not using a PEM encoding in ASCII.
+        return;
+    }
+    assert(buffer != NULL && *buffer != NULL);
+    assert(*len >= 0);
```
Comment (bartonjs) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:308]: The asserts seem like they should be before the first deferences :)

File: src/libraries/Native/Unix/Common/pal_utilities.h [hunk 2]
```diff
@@ -132,3 +132,5 @@ inline static int32_t SizeTToInt32(size_t value)
     assert(value <= INT_MAX);
     return (int32_t)value;
 }
+
+#define min(x, y) (x) < (y) ? (x) : (y)
```
Comment (bartonjs) [inline src/libraries/Native/Unix/Common/pal_utilities.h:136]: This is a double-evaluating version of this macro.  It's probably best to not define it at all.

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:308]: Yes they should :(
I added the first check after I wrote the assert.

File: src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c [hunk 3]
```diff
@@ -287,3 +294,58 @@ static int32_t PopulateByteArray(JNIEnv* env, jbyteArray source, uint8_t* dest,
...
+        if (lengthLocal == 0 || (iscntrl(*bufferLocal) && !isspace(*bufferLocal)))
+        {
+            // Either the buffer is empty or we encountered a control character that isn't whitespace.
+            // In the empty case, we didn't find the PEM header, so we can't skip to it.
+            // In the control character case, we know that this isn't explanatory info since that needs to
+            // all be printable or whitespace characters, not non-whitespace control characters.
+            return;
+        }
+        
+        if (memcmp(bufferLocal, pemHeader, min_size_t((size_t)lengthLocal, sizeof(pemHeader) - 1)) == 0)
```
Comment (bartonjs) [inline src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:337]: Hm.  Could you avoid the need for min_size_t if you just bailed earlier once `lengthLocal <= sizeof(pemHeader)` (vs `== 0`)?

The absolute minimum is something I don't really want to compute, but it's at least `-----BEGIN CERTIFICATE-----\n-----END CERTIFICATE-----`, which I count as 53 characters.  (The minimum value for a Base64 payload that could plausibly be a certificate is the part I don't want to do right now :smile:)

Comment (jkoritzinsky) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c:337]: I can add the earlier bailout.

Do we only need to handle the "BEGIN CERTIFICATE" case, or do we need to also be able to handle the other headers for PEM envelopes (like `X509 CRL`)?

Comment (bartonjs) [reply src/libraries/Native/Unix/System.Security.Cryptography.Native.Android/pal_x509.c]: Well, `X509 CRL` wouldn't be an appropriate match here, but `CMS` might, so I guess the minimum is only somewhere above 20 (hypens) + 2 (spaces) + 6 (2x "CMS".Length) + 5 (BEGIN) + 3 (END) => 36.

But using sizeof(pemHeader) is an obvious "the next part can't match", so it's good enough.

**Review by bartonjs — APPROVED** [review]

**Pull request merged**